velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp

```diff
@@ -0,0 +1,381 @@
+    }
+    // get metadata and initialize length
+    auto metadata = client_->GetObjectMetadata(bucket_, key_);
+    if (!metadata.ok()) {
+      checkGCSStatus(
+          metadata.status(),
+          "Failed to get metadata for GCS object",
+          bucket_,
+          key_);
+    }
+    length_ = (*metadata).size();
+    VELOX_CHECK_GE(length_, 0);
+  }
+
+  std::string_view pread(uint64_t offset, uint64_t length, void* buffer)
+      const override {
+    preadInternal(offset, length, static_cast<char*>(buffer));
+    return {static_cast<char*>(buffer), length};
+  }
+
+  std::string pread(uint64_t offset, uint64_t length) const override {
+    std::string result(length, 0);
+    char* position = result.data();
+    preadInternal(offset, length, position);
+    return result;
+  }
+
+  uint64_t preadv(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>>& buffers) const override {
+    // 'buffers' contains Ranges(data, size)  with some gaps (data = nullptr) in
+    // between. This call must populate the ranges (except gap ranges)
+    // sequentially starting from 'offset'. If a range pointer is nullptr, the
+    // data from stream of size range.size() will be skipped.
+    size_t length = 0;
+    for (const auto range : buffers) {
+      length += range.size();
+    }
+    std::string result(length, 0);
+    preadInternal(offset, length, static_cast<char*>(result.data()));
+    size_t resultOffset = 0;
+    for (auto range : buffers) {
+      if (range.data()) {
+        memcpy(range.data(), &(result.data()[resultOffset]), range.size());
+      }
+      resultOffset += range.size();
+    }
+    return length;
+  }
+
+  uint64_t size() const override {
+    return length_;
+  }
+
+  uint64_t memoryUsage() const override {
+    // TODO: Check if any buffers are being used by the GCS library
+    return sizeof(gcs::Client) + kUploadBufferSize + 2 * sizeof(std::string) +
+        sizeof(int64_t);
+  }
+
+  bool shouldCoalesce() const final {
+    return false;
+  }
+
+  std::string getName() const override {
+    return key_;
+  }
+
+  uint64_t getNaturalReadSize() const override {
+    return kUploadBufferSize;
+  }
+
+ private:
+  // The assumption here is that "position" has space for at least "length"
+  // bytes.
+  void preadInternal(uint64_t offset, uint64_t length, char* position) const {
+    gcs::ObjectReadStream stream = client_->ReadObject(
+        bucket_, key_, gcs::ReadRange(offset, offset + length));
+    if (!stream) {
+      checkGCSStatus(
+          stream.status(), "Failed to get GCS object", bucket_, key_);
+    }
+
+    stream.read(position, length);
+    if (!stream) {
+      checkGCSStatus(
+          stream.status(), "Failed to get read object", bucket_, key_);
+    }
+    bytesRead_ += length;
+  }
+
+  std::shared_ptr<gcs::Client> client_;
+  std::string bucket_;
+  std::string key_;
+  std::atomic<int64_t> length_ = -1;
+};
+
+} // namespace
+
+namespace filesystems {
+
+class GCSConfig {
```

Comment:
These config params should be moved to `HiveConfig` and documented. See PR for S3.
https://github.com/facebookincubator/velox/pull/4676/files

Reply:
Done.